sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/GpuHiveTableScanExec.scala

```diff
@@ -0,0 +1,501 @@
+   */
+  private def sendDriverMetrics(): Unit = {
+    driverMetrics.foreach(e => metrics(e._1).add(e._2))
+    val executionId = sparkContext.getLocalProperty(SQLExecution.EXECUTION_ID_KEY)
+    SQLMetrics.postDriverMetricUpdates(sparkContext, executionId,
+      metrics.filter(e => driverMetrics.contains(e._1)).values.toSeq)
+  }
+
+  private def buildReader(sqlConf: SQLConf,
+                          broadcastConf: Broadcast[SerializableConfiguration],
+                          rapidsConf: RapidsConf,
+                          dataSchema: StructType,
+                          partitionSchema: StructType,
+                          readSchema: StructType,
+                          options: Map[String, String]
+              ): PartitionedFile => Iterator[InternalRow] = {
+    val readerFactory = GpuHiveTextPartitionReaderFactory(
+      sqlConf = sqlConf,
+      broadcastConf = broadcastConf,
+      inputFileSchema = dataSchema,
+      partitionSchema = partitionSchema,
+      requestedOutputDataSchema = readSchema,
+      requestedAttributes = requestedAttributes,
+      maxReaderBatchSizeRows = rapidsConf.maxReadBatchSizeRows,
+      maxReaderBatchSizeBytes = rapidsConf.maxReadBatchSizeBytes,
+      metrics = allMetrics,
+      options = options
+    )
+
+    PartitionReaderIterator.buildReader(readerFactory)
+  }
+
+  /**
+   * Prune output projection to those columns that are to be read from
+   * the input file/buffer, in the same order as [[requestedAttributes]]
+   * Removes partition columns, and returns the resultant schema
+   * as a [[StructType]].
+   */
+  private def getRequestedOutputDataSchema(tableSchema: StructType,
+                                           partitionAttributes: Seq[Attribute],
+                                           requestedAttributes: Seq[Attribute]): StructType = {
+    // Read schema in the same order as requestedAttributes.
+    // Note: This might differ from the column order in `tableSchema`.
+    //       In fact, HiveTableScanExec specifies `requestedAttributes` alphabetically.
+    val partitionKeys: HashSet[String] = HashSet() ++ partitionAttributes.map(_.name)
+    val requestedCols = requestedAttributes.filter(a => !partitionKeys.contains(a.name))
+                                           .toList
+    val distinctColumns = requestedCols.distinct
+    val distinctFields  = distinctColumns.map(a => tableSchema.apply(a.name))
+    StructType(distinctFields)
+  }
+
+  private def createReadRDDForDirectories(readFile: PartitionedFile => Iterator[InternalRow],
+                                          directories: Seq[(URI, InternalRow)],
+                                          readSchema: StructType,
+                                          sparkSession: SparkSession,
+                                          hadoopConf: Configuration): RDD[InternalRow] = {
+    def isNonEmptyDataFile(f: FileStatus): Boolean = {
+      if (!f.isFile || f.getLen == 0) {
+        false
+      }
```

Comment:
nit: we usually have the else on the same line as the '}' before it. Is this code copied from some other place? if so then it is fine and we should probably keep it.

Reply:
This is my doing. I'll fix the formatting.